### ISSUE TITLE
Detect re-entrant substitute calls

### ISSUE THREAD

**Issue by dtchepak** [nsubstitute/NSubstitute.Analyzers#12]
Not sure how feasible this is but...

Call a substitute from within the `.Returns(...)` block of another substitute can cause issues:

https://github.com/nsubstitute/NSubstitute/issues/26
https://github.com/nsubstitute/NSubstitute/issues/357

Note that `sub.Blah().Returns(x => CreateFoo())` can be ok, whereas `sub.Blah().Returns(CreateFoo())` can cause the problem (when `CreateFoo` also sets `Returns()` values).

**Comment by dtchepak** [comment]
I'm just noting down potential cases for detection without thinking about feasibility. 😄 Feel free to ignore/close any suggestions!

**Comment by tpodolak** [comment]
I did an initial look into and I have some working PoC. Can you guys advise about what kind of message do you expect to see? How should I show to the user that given Returns setup calls (possible multiple) other Returns methods inside?

**Comment by dtchepak** [comment]
I think something like:

> `Returns()` is set with a method that itself calls `Returns`. This can cause problems with NSubstitute. Consider replacing with a lambda: `Returns(x => TheMethod())`.

I think the root method should be sufficient. What do you think?

**Comment by dtchepak** [comment]
It might just be sufficient to warn on `.Returns(SomeMethodCall())` and recommend `.Returns(x => SomeMethodCall())` instead? Might be easier with less performance impact?

**Comment by tpodolak** [comment]
The warning/diagnostic on the root call seems to be ok. I need one clarification though, putting aside potential performance problems, are all reentrant calls problematic from the perspective of NSubstitute or only the ones which are evaluated eagerly? Please take a look at the examples below and comment on them
````c#
substitute.Bar().Returns(ReturnThis); // no warning as invoking with delegate ?
substitute.Bar().Returns(MyMethod()); // no warning as MyMethod() returns a delegate?
substitute.Bar().Returns(ReturnThis()); // warning as invoking with result of function which calls Returns ?
substitute.Bar().Returns(ReturnForAnyArgsThis()); // warning as invoking with result of function which calls ReturnWithAnyArg ?
 substitute.Bar().Returns(WhenDoThis()); // warning as invoking with result of function which calls When..Do?

private int ReturnThis(CallInfo arg)
{
  Substitute.For<IFoo>().Bar().Returns(1);
  return 1;
}
private int ReturnThis()
{
  Substitute.For<IFoo>().Bar().Returns(1);
  return 1;
}

private int ReturnForAnyArgsThis()
{
  Substitute.For<IFoo>().Bar().ReturnForAnyArg(1);
  return 1;
}

private int WhenDoThis()
        {
            var @for = Substitute.For<IFoo>();
            @for.When(x => x.Bar()).Do(yy => { });
            return 1;
        }

 Func<CallInfo, int> MyMethod()
        {
            return ReturnThis;
        }
````
